modules/tftp/http_downloads.rb

```diff
@@ -0,0 +1,318 @@
+      end
+
+      def stop
+        @stop = true
+      end
+    end
+
+    attr_reader :status, :m
+
+    def initialize(dst_path, url, status = Status.new)
+      @status = status
+      @dst = dst_path
+      @url = URI.parse(url)
+      @downloader = nil
+      @m = Mutex.new
+    end
+
+    def is_local_copy_stale?(filepath)
+      return true unless File.exist?(filepath)
+      to_return = Concurrent::Promise.new { new_downloader.size }
+                                     .on_success { |result| result == 0 || result != File.size(filepath) }
+                                     .on_error { true }
+                                     .execute
+                                     .value(10) # don't block for too long on the check, fail it if we don't receive a response in 10 seconds
+      to_return.nil? ? true : to_return # Concurrent::Promise#value returns nil if the promise is in 'pending' state
+    end
+
+    def restart_download
+      Concurrent::Promise.new { new_downloader.restart }
+                         .on_error {|reason| attempt_restart(reason, 1) }
+                         .on_error {|reason| attempt_restart(reason, 2) }
+                         .on_error {|reason| attempt_restart(reason, 3) }
+    end
+
+    def download
+      Concurrent::Promise.new { new_downloader.start }
+                         .on_error {|reason| attempt_restart(reason, 1) }
+                         .on_error {|reason| attempt_restart(reason, 2) }
+                         .on_error {|reason| attempt_restart(reason, 3) }
+    end
+
+    def attempt_restart(reason, attempt_number)
+      case reason
+        when Errno::ECONNREFUSED
+        when Errno::ECONNRESET
+        when Errno::EHOSTUNREACH
+        when Errno::ETIMEDOUT
+        when ::Proxy::TFTP::HttpError
+          # retry on server error or http request timeout
+          if reason.code == 500 || reason.code == 408
+            sleep(attempt_number * ATTEMPT_DELAY * 3)
+            new_downloader.restart
+          elsif reason.code == 416
+            # no need to delay the download attempt -- the problem is with range header, not connection
+            new_downloader.start
+          else
+            raise reason
+          end
+        else
+          raise reason
+      end
+    end
+
+    def new_downloader
+      m.synchronize { @status = Status.new; @downloader = Downloader.new(@url, @dst, @status) }
+    end
+
+    def stop
+      m.synchronize { @downloader.stop }
+    end
+  end
+
+  class HttpDownloads
+    include ::Proxy::Log
+
+    CLEANUP_INTERVAL = 60*60*2 # 2 hours
+
+    # download_in_progress is used in tests only
+    attr_reader :m, :url_to_download, :id_to_download, :downloader_class, :download_in_progress
+
+    def initialize(dst_dir, downloader_class = HttpDownload)
+      @url_to_download = {}
+      @id_to_download = {}
+      @dst_dir = Pathname.new(dst_dir).cleanpath.to_s
+      @downloader_class = downloader_class
+      @m = Mutex.new
```

Comment:
Ditto. Add descriptive name to all mutexes in this patch with comments what is the purpose. This one looked like thread-safe hash at first sight (I'd use Concurrent::Hash instead in this case). But then I see huge lock in `do_download` method, please explain in the comment.